src/config/index.ts

```diff
@@ -519,11 +506,98 @@ export type AdapterConfigFromSettings<T extends SettingsMap> = {
 
-export type CustomSettingsType<T = SettingsMap> = Record<keyof T, Setting>
-export type EmptySettings = Record<string, never>
-export type SettingsMap = Record<string, Setting>
+export type CustomSettingsDefinition<T = SettingsDefinitionMap> = Record<keyof T, SettingDefinition>
+export type EmptySettingsDefinitionMap = Record<string, never>
+export type SettingsDefinitionMap = Record<string, SettingDefinition>
 export type ValidationErrorMessage = string | undefined
+
+/**
+ * TODO: Improve docs
+ * This class will hold the processed config type, and the basic settings
+ * The idea is that you can no longer use a straight object, but have to build a config
+ * Then in the generics we can simply pass the type of this, and it will hopefully allow for simpler generics
+ * name is placeholder
+ */
+export class AdapterConfig<T extends SettingsDefinitionMap = SettingsDefinitionMap> {
+  settings!: AdapterSettings<T>
+
+  constructor(
+    private settingsDefinition: T,
+    private options?: {
+      /** Map of overrides to the default config values for an Adapter */
+      envDefaultOverrides?: Partial<BaseAdapterSettings>
+
+      /** TODO: complete */
+      envVarsPrefix?: string
+    },
+  ) {}
+
+  initialize() {
+    this.settings = buildAdapterSettings({
+      customSettings: this.settingsDefinition,
+      overrides: this.options?.envDefaultOverrides,
+      envVarsPrefix: this.options?.envVarsPrefix,
+    })
+  }
+
+  validate(): void {
+    const validationErrors: string[] = []
+    Object.entries(BaseSettingsDefinition as SettingsDefinitionMap)
+      .concat(Object.entries(this.settingsDefinition || {}))
+      .forEach(([name, setting]) => {
+        validateSetting(
+          name,
+          (this.settings as Record<string, ValidSettingValue>)[name],
+          setting,
+          validationErrors,
+        )
+      })
+
+    if (validationErrors.length > 0) {
+      throw new Error(
+        `Validation failed for the following variables:\n ${validationErrors.join('\n')}`,
+      )
+    }
+  }
+
+  /**
+   * Creates a list of key/value pairs that need to be censored in the logs
+   * using the sensitive flag in the adapter config
+   */
+  buildCensorList() {
+    const censorList: CensorKeyValue[] = Object.entries(
+      BaseSettingsDefinition as SettingsDefinitionMap,
+    )
+      .concat(Object.entries((this.settingsDefinition as SettingsDefinitionMap) || {}))
+      .filter(
+        ([name, setting]) =>
+          setting &&
+          setting.type === 'string' &&
+          setting.sensitive &&
+          (this.settings as Record<string, ValidSettingValue>)[name],
+      )
+      .map(([name]) => ({
+        key: name,
+        // Escaping potential special characters in values before creating regex
+        value: new RegExp(
+          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+          ((this.settings as AdapterSettings)[name]! as string).replace(
+            /[-[\]{}()*+?.,\\^$|#\s]/g,
```

Comment:
Did you come up with this or pull it from somewhere?

Reply:
Not me I think, I moved this from somewhere else in the code